### ISSUE TITLE
Elaboration of let alternative option (6): Null checks on local.get

### ISSUE THREAD

**Issue by kripken** [WebAssembly/gc#187]
This is split off from the [general discussion of let alternatives](https://github.com/WebAssembly/function-references/issues/44), to not clutter things there.

Option (6) means this:

* Non-nullable types can be used normally with `local.set`s and `local.get`s, and a `local.get` of a null (i.e., a use of the default value, as a `local.set` cannot assign a null to a non-nullable local) will trap at runtime.

Adding null checks sounds like a step back, but it actually leads us two steps forward. For a **tl;dr** see the table at the bottom.

@rossberg had several concerns about this approach. I agree I was not clear enough before - sorry about that! Hopefully this post helps.

My specific claims are:

* Practically no runtime checks will occur in optimizing VMs.
* Only trivial work is required by optimizing VMs to achieve that.

(note that the claims are on optimizing VMs, not baseline; see later)

The claims are achieved by the following principle: Only emit a null check for a `local.get` if there is a path from the function entry to it which does not cross a `local.set`.

Before getting to examples, note that this is not reducible to the halting problem. All this requires is [use-def analysis](https://en.wikipedia.org/wiki/Use-define_chain), which is simpler than SSA. Every single optimizing compiler I am aware of uses SSA, and SSA provides use-def info automatically. Given use-def info, the VM can just look if the definition at the function entry (the default value) is used at each `local.get`. That is why I say this is trivial for optimizing VMs to do.

As a first example, consider a `let`-like pattern where a `get` is dominated by a `set`:

```wat
(func $foo (result (ref $X))
  (local $x (ref $X))
  (local.set $x (..))
  [..]
  (local.get $x) ;; no null check! :)
)
```

Dominance gives us the property we need, and there will be no null check. In other words, whenever a producer could emit a `let`, it could instead emit the code without a `let`, and we would still avoid null checks. So, at least in optimizing compilers, option (6) gives the same guarantees as `let` does.

But option (6) can do more than `let` as it can also handle phis:
```wat
(func $foo (result (ref $X))
  (local $x (ref $X))
  (if
    (..condition..)
    (block
      (..)
      (local.set $x (..value1))
      (..)
    )
    (block
      (..)
      (local.set $x (..value2))
      (..)
    )
  )
  (local.get $x) ;; no null check! :)
)
```
This if-else sets a value somewhere in each of its branches, and so the `get` has no path from the default.

Now, there is some risk here, of a compiler finding "clever" ways to emit the code, such as
```wat
;; bad example 1
(func $foo (result (ref $X))
  (local $x (ref $X))
  (if
    (..complex condition1..)
    (local.set $x (..))
  )
  (if
    (..complex condition2..)
    (local.set $x (..))
  )
  (local.get $x) ;; null check :(
)
```
Say that the producer knows that either condition1 or 2 will be true. Then logically no null check is needed, but the wasm VM will emit one. Another case that cannot be optimized is
```wat
;; bad example 2
(func $foo (result (ref $X))
  (local $x (ref $X))
  (if
    (i32.const 1) ;; always taken
    (local.set $x (..))
  )
  (local.get $x) ;; null check :(
)
```
Unless the optimizing compiler understands that the if is always taken, it will do a null check. And here is a third case:
```wat
;; bad example 3
(func $foo (result (ref $X))
  (local $x (ref $X))
  (if
    (..complex condition that may or may not be true..)
    (local.set $x (..))
  )
  (local.get $x) ;; null check :(
)
```
This null check is actually logically necessary. It represents an invalid program (assuming it was compiled from something with a non-nullable type there).

Are these a problem? Bad example 3 is just a compiler bug and can be ignored. But 1 and 2 are worth discussing. First, such issues can be avoided by using nullable types for phis, and only using non-nullable ones where `let` would have been used, as mentioned before. So we don't need to do anything more than what `let` allows, if we don't want to.

But avoiding null checks on phis sounds useful, and at least in all the compiler output I've seen - mainly LLVM, but not only - I think things would work fine. The "clever pattern" in bad example 1, and things like unoptimized code as in bad example 2 are not realistic code patterns. In general phis are emitted in a straightforward way of assigning the values right before the merge. When doing it that way, the only risk is a compiler correctness bug.

We can also provide tools to help with this. A trivial Binaryen pass could log if there are any places where a null check would be emitted for a `local.get` (trivial, because like in a wasm VM we'd just check the results of our use-def analysis).

What happens if in rare cases a null check "slips through"? This would be on something that `let` cannot handle anyhow, so it is not a regression compared to that. And if it's not a compiler bug, maybe it is a rare code pattern that we can't expect VMs to understand and optimize - a null check may actually be necessary there. But even so, there is some benefit to option (6) here, in that we can just write code with a non-nullable type, instead of using a nullable one plus some added `ref.as_non_null`s.

Note that the option of using nullable types plus `ref.as_non_null` is just as optimizable as option (6), using the same principles:
```wat
(func $foo (result (ref $X))
  (local $x (ref null $X)) ;; a nullable type
  ;; a phi as before
  (if
    (..condition..)
    (local.set $x (..non-null value1))
    (local.set $x (..non-null value2))
  )
  (ref.as_non_null ;; no null check! :)
    (local.get $x)
  )
)
```
However, as just mentioned option (6) allows more compact binaries as we can avoid `ref.as_non_null`s.

The main downside of option (6) is that we will have null checks in baseline compilers. Here is how I'd summarize the tradeoffs:

|                           | let  | option (6)  |
|---------------------------|--------|-------------|
| optimizing compiler: `let` pattern, `set` dominating `get`s |  **faster**      |    **faster**         |
| optimizing compiler: control flow merge of `set`s |  slower      |    **faster** (with slight risk)        |
| baseline compiler  |  **faster**       |    slower         |
| startup speed   |  slower     |    **faster**      |
| VM complexity      |  larger     |    **smaller**      |

(**bold** are the better results)

* Compiler output speed of "faster" means no null checks, "slower" means the opposite. Startup speed of "faster" indicates a smaller binary size (`let` leads to larger binary sizes due to the `let` blocks and variable declarations, etc., see the original issue in the link at the top; note that option (6) adds no bytes at all.)
* See discussion of VM complexity of `let` in the original issue. Note that option (6) adds almost no complexity either to the spec or to VMs: it just adds null checks to a single instruction.

In summary, option (6) is simple, allows for the most guaranteed speed, and the fastest startup. The cost is baseline speed. Overall it looks like a good tradeoff to me :smiley:

**Comment by rossberg** [comment]
Thanks for writing this up. Just a few notes for now.

> Before getting to examples, note that this is not reducible to the halting problem. All this requires is use-def analysis, which is simpler than SSA.

Just to make sure we are on the same page, the problem _does_ reduce to the halting problem -- a use-def analysis is merely a decidable approximation, and your later examples in fact demonstrate that it is necessarily incomplete. Of course, there is an argument to be made that this would be totally sufficient in practice, and I'd agree.

It's still not entirely predictable, though. Your counter examples use "complex conditions", but what exactly that is would be up to an engine to decide. One would probably expect an engine to handle, e.g., `if (true)` specially, but there is a wide range of possible expectations beyond that.

Also, the performance analysis only considers time. One downside of dead zones is that they can also take up additional space -- not just in code but in memory -- if you cannot optimise them. It's not always possible to represent the uninit status by a sentinel value, e.g., consider numeric types. The induced overhead would particularly affect baseline implementations, despite those being the preferred choice in domains with space constraints.

As an aside, @titzer would probably complain that the table doesn't consider the interpreter use case, which would also be slower. And code verification -- increasingly important in Wasm's niches -- becomes harder as well.

Finally, I still wonder whether this approach isn't practically equivalent to simply specifying that locals cannot have non-nullable reference type. Isn't the practical effect almost the same, just with an extra mechanism?

**Comment by jakobkummerow** [comment]
I generally like this approach.

Regarding baseline compiler performance: I'd say "it depends". If the baseline compiler is implemented such that `let` has to physically move the current stack frame's contents (existing locals + value stack) in order to make room for a new local at index 0, then doing a few null checks instead might not be any slower than that.

> I still wonder whether this approach isn't practically equivalent to simply specifying that locals cannot have non-nullable reference type. Isn't the practical effect almost the same, just with an extra mechanism?

Yes, the difference between this and just using nullable locals seems small. Optimizing engines will likely optimize away the same set of null-checks. I guess the biggest difference is that having non-nullable locals and trapping `local.get` avoids the need for an explicit `ref.as_non_null` that would otherwise be required before an operation that requires an input that's statically typed to be non-null -- and I don't currently see any such instructions in the list. So we could look at spec simplicity (=lack of special cases) as a criterion, but that seems to be a wash as well: we either have to specify an extra note on `local.get` ("traps on non-nullable locals when the local is null") or on locals ("cannot have non-defaultable types").

**Comment by rossberg** [comment]
@jakobkummerow:
> > I still wonder whether this approach isn't practically equivalent to simply specifying that locals cannot have non-nullable reference type. Isn't the practical effect almost the same, just with an extra mechanism?
> 
> Yes, the difference between this and just using nullable locals seems small. Optimizing engines will likely optimize away the same set of null-checks. I guess the biggest difference is that having non-nullable locals and trapping `local.get` avoids the need for an explicit `ref.as_non_null` that would otherwise be required before an operation that requires an input that's statically typed to be non-null -- and I don't currently see any such instructions in the list.

Right, all current instructions allow null anyway and do the check if necessary. The only place where you'd need an explicit ref.as_non_null would be when passing it somewhere else expecting non-null type.

> So we could look at spec simplicity (=lack of special cases) as a criterion, but that seems to be a wash as well: we either have to specify an extra note on `local.get` ("traps on non-nullable locals when the local is null") or on locals ("cannot have non-defaultable types").

The former involves new run-time behaviours, the latter merely a validation-time restriction. Moreover, the former would also (rather uselessly) have to handle non-reference types, which are more complicated to implement. Unless you want to prohibit those, but then you have a similar static restriction in both cases, and the balance shifts even more towards the latter.

**Comment by kripken** [comment]
@rossberg 

> It's still not entirely predictable, though. Your counter examples use "complex conditions", but what exactly that is would be up to an engine to decide.

Yes. But to be clear, my claim is that even if the engine ignores **all** complex conditions - if all it does is use the simple principle mentioned - this would be enough for the great majority of cases which are (1) simple dominance (what `let` can do) and (2) phis. The example of "complex conditions" was meant to show something that it is reasonable to not handle. In particular I think it's fine if VMs do **not** handle `if (true)`. We can depend on producers to emit good code.

I did a quick measurement to add to the logical arguments now, compiling musl libc and libc++. There are 79,120 `local.get`s and only 33 cases in which the default value is used. The cases where it is used appear to be in large functions with irreducible control flow. This is obviously not a fully proper measurement - maybe things will be different with non-nullable reference types for some reason - but it reflects what I've seen, that producers almost always assign values on all paths to a `local.get`.

> Also, the performance analysis only considers time. One downside of dead zones is that they can also take up additional space -- not just in code but in memory -- if you cannot optimise them. It's not always possible to represent the uninit status by a sentinel value, e.g., consider numeric types.

Sorry, I don't follow this. What is a "dead zone", and what is "unint"?

> the table doesn't consider the interpreter use case

Good point, yes - anywhere I've said "baseline" I should have said "non-optimizing" or "baseline or interpreter".

@jakobkummerow 

> the difference between this and just using nullable locals seems small [..] I guess the biggest difference is [..] trapping local.get avoids the need for an explicit ref.as_non_null

Yes, exactly.

Aside from the spec and VM changes, there is also a "normative" difference of being focused on optimizing for the principle mentioned. That is, if we agree this approach makes sense, we would do work on the toolchain side to ensure we avoid paths from the default as much as possible, make tools as I mentioned to help find such issues, investigate the 33 cases mentioned earlier on musl/libc++, etc. (We can also do all that without option (6), but it makes sense together.)

**Comment by RossTate** [comment]
Ah, I understand your intent better now. Thanks for elaborating! Now I'm wondering how this would be intended to work with an imported type whose upper bound is non-nullable but which might still have a null value?

**Comment by tlively** [comment]
>> So we could look at spec simplicity (=lack of special cases) as a criterion, but that seems to be a wash as well: we either have to specify an extra note on local.get ("traps on non-nullable locals when the local is null") or on locals ("cannot have non-defaultable types").
>
> The former involves new run-time behaviours, the latter merely a validation-time restriction. Moreover, the former would also (rather uselessly) have to handle non-reference types, which are more complicated to implement.

Why is this more complicated to implement? I believe the intent here (and @kripken, correct me if I'm wrong) is that only gets of locals with non-defaultable types would trap, so gets of uninitialized RTT locals would trap, but gets of uninitialized numeric types would not.

**Comment by tlively** [comment]
> Ah, I understand your intent better now. Thanks for elaborating! Now I'm wondering how this would be intended to work with an imported type whose upper bound is non-nullable but which might still have a null value?

Nullable types are not subtypes of their corresponding non-nullable types (it goes the other way around), so I don't think it would be possible for such an imported type to have a null value.

**Comment by kripken** [comment]
@tlively 

> I believe the intent here (and @kripken, correct me if I'm wrong) is that only gets of locals with non-defaultable types would trap, so gets of uninitialized RTT locals would trap, but gets of uninitialized numeric types would not.

Yes, that's correct. I was only thinking of reference types here myself. For numeric types, we can't change the current behavior, it would break existing code that depends on getting a 0 that way.

**Comment by RossTate** [comment]
So what is the behavior of `local.get` of an imported type?

**Comment by tlively** [comment]
@RossTate I believe it would trap on uninitialized values if the imported type was not defaultable.

**Comment by RossTate** [comment]
Thanks for the clarification. Unfortunately that's problematic for type-erased polymorphism, as behavior would change depending on what the (erased) type parameter is.

**Comment by tlively** [comment]
I don't understand what the problem would be. In the absence of compiler bugs, local.get of uninitialized locals should never happen, so it shouldn't matter what its specified behavior is.

**Comment by kripken** [comment]
To add to @tlively 's point: A `local.get` of a null will trap depending on the value at runtime, not the type.

If the type is non-nullable then it prevents the possibility entirely, of course, but if you have all the same values but change the type to be nullable then there would be no observable behavior change.

You would need to change the type *and* change the runtime values to get a behavior change.

**Comment by RossTate** [comment]
If I have a local of an imported type that I initialize on one branch, but not the other, what happens when I do `local.get` after the two branches merge?

**Comment by tlively** [comment]
Assuming the type import is non-defaultable, the `local.get` would have to trap if the non-initializing branch had been taken and succeed otherwise.

**Comment by RossTate** [comment]
The issue is that you do not know if the type is defaultable or not. That is, it can have a default value, but you do not know it has a default value. So when you do `local.get`, and it is the default value, is that because it was not initialized, or is that because it was initialized (so through some function call) but it happened to be initialized to the default value?

**Comment by kripken** [comment]
@RossTate 

I see there is discussion about nullability and type imports: https://github.com/WebAssembly/proposal-type-imports/issues/11 where it is proposed to annotate type imports as having a default or not.

That makes sense to me, because if you don't know if an imported type has a default or not, how can you know if you can use it as a local or not? That is, not knowing defaultability seems to have bigger implications than option (6). But this is maybe a topic for that proposal.

**Comment by RossTate** [comment]
That issue suggests having a "has-default-value" so that, for example, people can use locals without `let`. I agree that would be helpful. But what I'm asking about is what if an import has no such constraint so that it can represent either nullable or non-nullable types? That doesn't mean it doesn't have a default value; it just means the importing module does not know whether or not it has a default value.

So what should the program I described do (supposing it takes the uninitialized path) in the following circumstances:
* the type import is constrained to be non-defaultable (@tlively already answered this, indicating it should trap)
* the type import is constrained to be defaultable
* the type import is not constrained either way, and is instantiated with a non-defaultable type
* the type import is not constrained either way, and is instantiated with a defaultable type

**Comment by kripken** [comment]
@RossTate 

I am guessing that a type import of type X behaves the same as if the module instead had X in all those places instead of importing it. Is that correct?

* If so, then the answer is straightforward, a non-nullable type will trap on a `local.get` of the default value, and that's the case whether the type was imported or not, and whether or not the imported type was marked as defaultable (so, in your list, the answers are "trap, not trap, trap, not trap").
* If not, then I have a lot more questions about type imports :smile: but those are probably out of scope for this issue.

**Comment by RossTate** [comment]
Your intuitions regarding type imports are correct 😄 But notice that you gave two different answers for the same program. That means the semantics doesn't support type erasure&mdash;the engine has to know whether the imported type is defaultable or not to determine how to compile the program. Now for type imports, erasure isn't critical (because instantiation is infrequent), but for polymorphism, erasure is critical (because function calls are frequent)&mdash;though not strictly necessary as we could go the .NET route. That's why I was saying that the (clarified) design about is problematic for type-erased polymorphism. (That doesn't mean it's not a viable design; it's just a tradeoff to be aware of.)

Hopefully that was clearer? Apologies for the back and forth; I just figured a more interactive back-and-forth would be a better means of explaining the concern.

**Comment by kripken** [comment]
@RossTate 

Makes sense, thanks - yes, there are a bunch of tradeoffs here.

It is also possible as others have said to not add the null checks of option (6) but to otherwise focus on the optimization route described in this issue. That would have different tradeoffs - larger code size (more `ref.as_non_null`), but would not have this issue.

I don't have a strong opinion between those myself.

**Comment by conrad-watt** [comment]
If the difference is mostly otherwise a wash, my preference would be to preserve the idea that `local.get/set` are always pure assignments. Conceptually it makes more sense to me that the `ref.*` instructions should be responsible for null checks.

**Comment by kripken** [comment]
I think I lean towards agreeing with @RossTate and @conrad-watt that adding traps on `local.get` may not be worth it, since the main downside is code size (and the main increase is that each `local.get` would be followed by a `ref.as_non_null` - that should at least gzip very well, and in theory we can even consider a single instruction for it if data justifies that).

Given that, this proposal could instead be:

* Focus on optimizing according to the principle stated at the top, but reformulated now as `if all paths to a local.get of a nullable type are from local.sets of non-nullable values, note that fact internally, and use it to avoid null checks on uses`. I.e., optimizing VMs would be expected to apply use-def analysis to achieve that (trivial, as mentioned earlier), and tools would be expected to emit it as much as possible.
  * In particular, get useful tools for diagnosing violations of the principle. (There is somewhat more work here on the tools side compared to the original proposal in this issue, since we do not mark the intention of something being non-null using its type, but it's still possible.)
* Measure the performance and size, and consider additional instructions and/or trapping on local.get later (both would be backwards compatible).

**Comment by RossTate** [comment]
Can you clarify how locals of non-defaultable types are handled in that revision? It sounds like you still expect to have them, but I'm not sure how they're initialized.

**Comment by kripken** [comment]
@RossTate Sorry, I wasn't clear enough, then. There are no non-defaultable locals in the variant without traps on `local.get`. Instead the optimization principle is used on defaultable ones. Example:
```wat
(func $foo (result (ref $X))
  (local $x (ref null $X)) ;; a nullable type
  ;; a phi
  (if
    (..condition..)
    (local.set $x (..non-null value1))
    (local.set $x (..non-null value2))
  )
  (ref.as_non_null ;; 2. use the mark => no null check! :)
    (local.get $x) ;; 1. all sets assign a non-null value, mark as such
  )
)
```
Note that this just requires simple use-def info, and not a flow algorithm or anything like that (the only extra work compared to the original description is that before we inferred whether a null check was needed on a `local.get` itself, and now we get the same information and use it where the `local.get` is used).

**Comment by RossTate** [comment]
I understand the rely-on-optimization perspective (I also expressed the opinion that non-nullable types would not be worth the hassle at the moment), but I don't see how this provides an alternative to `let` (per WebAssembly/function-references#44) at this point. Can you clarify with an example using non-defaultable locals?

**Comment by conrad-watt** [comment]
@RossTate with this proposed solution, locals are statically forbidden from having non-nullable types, and there is no `let`. This means that any non-nullable type on the stack must be upcasted to a nullable type when stored into a local, and later downcasted with `ref.as_non_null` if a subsequent use is required in its non-nullable form.

The point @rossberg made earlier is that this is broadly the same as allowing non-nullable locals with a dynamic initialisation check, and is subject to the same use-def/flow reasoning to eliminate `ref.as_non_null`.

**Comment by tlively** [comment]
I think the distinction between this idea and simply not having non-nullable types is that this idea still allows for declaring functions with non-nullable parameters and results. That's certainly less useful than also having non-nullable locals, but it doesn't require any other new mechanism like `let`.

**Comment by conrad-watt** [comment]
Certainly one would still expect non-nullable types to be more efficient if they're either used directly from the stack, or if upcasts due to `local.set` are obvious enough that engine optimisations can eliminate subsequent downcasts/null checks.

This approach would probably mean there's a stronger case for stack operations like `pick`, but this can be discussed separately since the only benefit would be potentially eliminating some null checks that the engine couldn't reason about.

**Comment by RossTate** [comment]
Ohh, I missed a key "no" in @kripken's comment! Sorry for the confusion.

That sounds fine to me; thanks for the clarification. Important implication is that every non-defaultable type will need a defaultable counterpart to be reasonably usable by generators. That in particular applies to type imports and `rtt`s (per #175). For type imports, that gets a little complicated, but maybe still manageable?

**Comment by kripken** [comment]
@conrad-watt 

> one could make the point that this reduces the value of non-nullable types, depending on how confident one is in the ability of engines to eliminate ref.as_non_null and similar implicit checks in other ref instructions.

That is the key question, yes. The main idea in this issue is that we can be confident - if we work together on optimizing VMs and tools, using the simple principle mentioned earlier. Both reasoning on dominance and phis supports that, [and some data](https://github.com/WebAssembly/gc/issues/187#issuecomment-769158006). I believe only rare cases of irreducible code will escape an optimized wasm file in an optimizing wasm VM.

It is possible something in that reasoning is wrong! But I've not seen that yet.

**Comment by rossberg** [comment]
(Oops, I lost this thread.)

@kripken:
> > Also, the performance analysis only considers time. One downside of dead zones is that they can also take up additional space -- not just in code but in memory -- if you cannot optimise them. It's not always possible to represent the uninit status by a sentinel value, e.g., consider numeric types.
> 
> Sorry, I don't follow this. What is a "dead zone", and what is "unint"?

Dead zone as in JS's lexical scopes: a program range where a variable is in scope but reading it causes an exception.

Sorry, uninit = uninitialised. If you want a clean semantics for let variables, then the same access restrictions ought to apply to all types. For reference types, an engine can typically use some sentinel reference to distinguish this state. But for numeric types, no such values are available, so they'd need to allocate extra flag bits somewhere.

Unless we want a semantics for let variables that depends on their type, such that e.g. dead zones only apply to (non-null?) reference types. But that would be rather hacky, and begs the question even more why to introduce such a let in the first place instead of just enforcing nullable refs.

@others:
There is some basic confusion regarding type imports in this thread. The type algebra carefully distinguishes between _heap type_ definitions and _value types_. A value type can be formed as a reference to a heap type, and can either allow or disallow null. Defaultability hence is a property of value types, not heap types. Type definitions and imports on the other hand always denote heap types, at least for now (otherwise compilation before linking wouldn't work), so defaultability is completely orthogonal to them.

(Some folks have asked for im/export of value types, but to clarify again: that is a separate feature and a _much_ bigger addition to the language. It would require first introducing a new, staged compilation and linking model with premodules, preimports, and so on. That is possible, but far outside the scope of the current type import proposal.)

**Comment by kripken** [comment]
@rossberg 

I'm a little confused by your reply, statements like `If you want a clean semantics for let variables` and `begs the question even more why to introduce such a let in the first place` - what is proposed here is an alternative to `let`, so there is no proposal for `let` semantics here?

Or do you mean that trapping on `local.get` introduces "`let`-like" semantics in a sense?

In any case, [more recent discussion](https://github.com/WebAssembly/gc/issues/187#issuecomment-769983081) in this issue is heading in the direction of not having `local.get` trap, and instead rely on the optimization principle mentioned here as a convention between tools and VMs to ensure non-nullable operations are optimized. In that case there are no non-nullable locals, no dead zones, etc., but optimizing VM performance should still be maximal (at the cost of some code size).

**Comment by rossberg** [comment]
@kripken, oops, sorry, that was supposed to say something like "semantics for local variables without initialisation". I was assuming that they would somehow have to be declared as such explicitly, in a way that is independent from typing.

**Comment by kripken** [comment]
@rossberg The original idea here was like that, yes. But the option that seems to make more sense is to not allow non-nullable locals, as it still allows for the same optimizations but without adding complexity (see discussion above).

**Comment by kripken** [comment]
Yes, let's close this. The benefit to this approach is that no annotations are needed, but meanwhile @jakobkummerow and @askeksa-google have shown that not having annotations is viable in a better way, that does not have downsides to baseline compilers like this idea.

**Comment by kripken** [comment]
Actually, let me reopen this, sorry for the noise. Comparing 6 (traps) to option 1c (no annotations, infer during validation), I think combining the two is better than either:

* A key insight from @jakobkummerow and @askeksa-google on 1c is that tracking which locals are set is easy to compute, with or without annotations (though we still need to check if 1b is even faster). We can use the same VM implementation in option 6, that is, the spec says `local.get` traps on a null, but the VM can track local sets and remove relevant null checks exactly as with 1c. That would give us the same performance in baseline compilers as 1c, that is, option 6 no longer has a downside there.
  * In other words, when I said in the first comment "`Only trivial work is required by optimizing VMs to achieve [no null checks]`", I was too pessimistic: the same is true for baseline as well (and even interpreters, but they'd need to save the inferred data I guess).
* In option 6 that computation is **optional**. Since it isn't necessary for validation, a VM could simply not do it at all (and emit null checks). That's simpler, and it might be an interesting tradeoff on speed: faster validation / slower baseline / no change to optimizing tier.
* Option 6 is **the simplest to spec**: we don't need to define an inference validation algorithm at all, and we don't need to spec new block annotations. We just spec that `local.get` of a non-nullable local traps on null, similar to how `struct.get` etc. trap on null.

OTOH, the main advantage of 1c over 6, I think, is that on the rare "bad" cases 1c fails on validation where 6 has a null check at runtime. Erroring early has benefits, but as mentioned in the top comment there are downsides too, and optimizing for rare and poor code shouldn't be our top priority I think.

Very possibly I'm missing something though?

Note btw that in the later parts of the above discussion we were leaning towards just using `ref.as_non_null` instead of traps. However, the new data favors traps as well: as shown by @jakobkummerow for 1c, avoiding `ref.as_non_null` leads to smaller binaries and faster startup even with a small validation penalty due to 1c. Since with traps we can optionally avoid that penalty as well, startup could be even faster.

**Comment by RossTate** [comment]
Per the observation in https://github.com/WebAssembly/function-references/issues/44#issuecomment-1156495707, I believe this is the only still-active contender for which WebAssembly functions remain decomposable even with the current EH design. I believe that means that it is the only option that can make a formal claim to being "consistent" with WebAssembly's existing design.

**Comment by titzer** [comment]
It's worth noting that nullchecks on `local.get` penalize all local accesses on interpreters.

**Comment by kripken** [comment]
@titzer 

Yes, that may be a downside of this option. One thing discussed (looks like not here, so must have been in the other thread) was `local.get_as_non_null` or such, that is, a new instruction, which would avoid any change at all to nullable (or non-reference) locals. Also, an interpreter that can save the annotations from the validation-like step can also avoid a slowdown, as mentioned in my last comment, but of course that just helps some interpreters.

(I think there were some other options proposed last time we discussed this, but it's been a while...)

**Comment by RossTate** [comment]
Oh, yes, to correct myself above: the designs where all local types remain defaultable also have decomposable functions even with EH, though only blocks with all defaultable types can be pulled out into a separate function.

**Comment by jakobkummerow** [comment]
Allowing engines the option to verify locals initialization status on access via `local.get` is an interesting idea to accommodate implementation flexibility (@tebbi has also suggested it in offline conversation), but I see one significant problem with it: it's necessarily 100% precise, whereas for the options "1b" and "1c" discussed in the other issue, we would (very likely) make certain simplifying restrictions. As a specific example, if we define that "`try..catch is like if..else`", then an engine that chooses the "trap on `local.get`" approach would accept `(try (local.set $x) (throw ...) catch ...) (local.get $x)` as valid (where the `catch`-block doesn't initialize `$x`, but the `local.set` in the `try`-block is always executed before the first thing that can possibly throw), whereas an engine that chooses to check locals as part of validation would reject it based on the simplified semantics.

(It's probably possible to address this, but then the "trap on `local.get`" approach would mean entertaining a _separate_ array of initialization status information, as opposed to being able to use internal sentinels for the local's value, which would probably make it an unattractive implementation option.)

**Comment by RossTate** [comment]
I think the intent of this design is that engines never reject programs (statically) based on (un-)initialization of locals. Rather, engines are expected to use `null` to encode uninitialization of non-null-types locals and fail when `local.get` would (dynamically) return a `null` value. As you've demonstrated with your experiments, engines can (choose to) easily eliminate the dynamic checks for initialization in any of the cases that any approach in the other thread would be able to statically validate. On the other hand, the approach here would also be able to accommodate (using fairly inexpensive dynamic techniques) initialization patterns that the static approaches fail to recognize (say, due to EH).

**Comment by kripken** [comment]
@jakobkummerow Good points. Yes, it seems risky to allow "mixing" of the two strategies of either trapping at runtime or erroring at validation.

If I wasn't clear enough before, in my last comment I meant what @RossTate said: that the `trap` option would validate your `try-catch` example, and engines could, optionally and as an internal optimization, run a 1c-like analysis to remove null checks. Engines doing so would *not* remove the `try-catch` null check in your example, and so that might be a remaining null check (but some optimizing compilers might remove it?). I think the impact there is low: we have reason to believe 1c handles the great majority of cases, because even joins (which 1c handles) [are rare](https://github.com/WebAssembly/function-references/issues/44#issuecomment-953744506), and this is likely more rare than them, if only because `try` itself is fairly rare. (And if this was not rare, it would be a problem for 1c (and 1b) as well, as it would mean common code patterns could not even validate - but again, I expect those are rare.)

**Comment by tlively** [comment]
We settled on option 1a, so I'll go ahead and close this.
